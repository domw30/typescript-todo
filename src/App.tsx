import React, { Fragment, useState } from 'react';
import { Todo } from './interfaces/Todo';

type FormElem = React.FormEvent<HTMLFormElement>;

function App(): JSX.Element {
  const [value, setValue] = useState<string>('');
  const [todos, setTodos] = useState<Todo[]>([]); // useState is expecting an array of interface Todos

  const handleSubmit = (e: FormElem): void => {
    e.preventDefault(); // prevents the form from doing a refresh.
    addTodo(value);
    setValue(''); // Turns value in input to empty string.
  };

  const addTodo = (text: string): void => {
    const newTodos: Todo[] = [...todos, { text, complete: false }]; // spread operator brings in previous toods. new array for new todos
    setTodos(newTodos);
  };

  // completeTodo takes an argument of index dos array called, newTodos
  const completeTodo = (index: number): void => {
    const newTodos: Todo[] = todos;
    newTodos[index].complete = !newTodos[index].complete;
    // toggle the complete property in the object of our todo by using index to locate it.
    setTodos(newTodos);
    // set object as our new todo array with the setTodos function.
  };

  const removeTodo = (index: number): void => {
    const newTodos: Todo[] = todos;
    newTodos.splice(index, 1);
    setTodos(newTodos);
  };

  console.log(todos);

  return (
    <Fragment>
      <h1>Todo List</h1>
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          value={value}
          onChange={e => setValue(e.target.value)}
          required
        />
        <button type="submit">Add Todo</button>
      </form>
      <section>
        {todos.map((todo: Todo, index: number) => (
          <div key={index} style={{ display: 'flex' }}>
            <div
              style={{ textDecoration: todo.complete ? 'line-through' : '' }}
            >
              {todo.text}
            </div>
            <button type="button" onClick={() => completeTodo(index)}>
              {todo.complete ? 'Incomplete' : 'Complete'}
            </button>
            <button onClick={() => removeTodo(index)}>X</button>
          </div>
        ))}
      </section>
    </Fragment>
  );
}

export default App;
