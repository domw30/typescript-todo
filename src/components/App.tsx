import React, { Fragment, useState, useRef } from 'react';
import { Todo } from '../interfaces/todo';
import { GlobalStyles } from '../styles/global-styles';
import { DeleteIcon, CheckIcon, EditIcon } from '../styles/icons';
import {
  TodoListWrapper,
  TodoListItem,
  CompleteButton,
  DeleteButton,
  TodoList,
  EditButton,
} from '../styles/styles';
import Header from './header';
import TodoForm from './todo-form';

function App(): JSX.Element {
  const [inputValue, setValue] = useState<string>('');
  const [todos, setTodos] = useState<Todo[]>([]);

  const inputFocus = useRef<HTMLInputElement>(null);

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>): void => {
    e.preventDefault();
    addTodo(inputValue);
    setValue('');
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    e.preventDefault();
    setValue(e.target.value);
  };

  const addTodo = (text: string): void => {
    const newTodos: Todo[] = [
      ...todos,
      { text, complete: false, edited: false },
    ];
    setTodos(newTodos);
  };

  const completeTodo = (index: number): void => {
    const newTodos: Todo[] = [...todos];
    newTodos[index].complete = !newTodos[index].complete;
    setTodos(newTodos);
  };

  const removeTodo = (index: number): void => {
    const newTodos: Todo[] = [...todos];
    newTodos.splice(index, 1);
    setTodos(newTodos);
  };

  const editTodo = (index: number): void => {
    const newTodos = [...todos];
    newTodos[index].edited = !newTodos[index].edited;
    if (true === newTodos[index].edited && inputFocus.current) {
      inputFocus.current.focus();
    }
    console.log(todos);
    setTodos(newTodos);
  };
  return (
    <Fragment>
      <GlobalStyles />
      <Header />
      <TodoForm
        onSubmit={handleSubmit}
        onChange={handleChange}
        value={inputValue}
      />
      <TodoListWrapper>
        {todos
          .map((todo: Todo, index: number) => (
            <TodoList key={index}>
              <TodoListItem
                data-type="todo-item"
                complete={todo.complete}
                edited={todo.edited}
              >
                {todo.text}
              </TodoListItem>
              <CompleteButton
                type="button"
                data-type="check-button"
                onClick={(): void => completeTodo(index)}
              >
                <CheckIcon />
              </CompleteButton>
              <DeleteButton
                type="button"
                data-type="delete-button"
                onClick={(): void => removeTodo(index)}
              >
                <DeleteIcon />
              </DeleteButton>
              <EditButton
                type="button"
                data-type="edit-button"
                onClick={(): void => editTodo(index)}
              >
                <EditIcon />
              </EditButton>
              <input ref={inputFocus} type="text" onChange={handleChange} />
              <button>Save</button>
              <button>Cancel</button>
            </TodoList>
          ))
          .reverse()}{' '}
        }
      </TodoListWrapper>
    </Fragment>
  );
}

export default App;
