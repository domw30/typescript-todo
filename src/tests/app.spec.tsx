import React from 'react';
import { shallow } from 'enzyme';
import App from '../components/app';
import renderer from 'react-test-renderer';
import Header from '../components/app';
import TodoForm from '../components/app';

describe('First React component test with Enzyme', () => {
  it('renders without crashing', () => {
    shallow(<App />);
    // shallow assertion method. Igonres any child components. E.g. Header will be ignored.
    // A typed of unit test for React, designed to test one function.
    // smoke test. Verifies a component renders without crashing/throwing.
  });
});

describe('First Snapshot test', () => {
  it('app should render correctly', () => {
    const app = renderer.create(<App />).toJSON();
    expect(app).toMatchSnapshot();
  });

  it('should render header correctly', () => {
    const header = shallow(<Header />);
    expect(header).toMatchSnapshot();
  });

  it('should render TodoForm correctly', () => {
    const todoForm = shallow(<TodoForm />);
    expect(todoForm).toMatchSnapshot();
  });
});
