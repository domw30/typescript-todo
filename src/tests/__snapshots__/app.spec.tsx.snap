// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`First Snapshot test app should render correctly 1`] = `
Array [
  <header
    className="sc-bxivhb bUZGRr"
  >
    <h1
      className="sc-ifAKCX hVeHwL"
    >
      Todo List
    </h1>
  </header>,
  <form
    className="sc-EHOje iJkfhn"
    onSubmit={[Function]}
  >
    <input
      className="sc-htoDjs QZhGa"
      onChange={[Function]}
      placeholder="Enter Todo"
      required={true}
      type="text"
      value=""
    />
    <button
      className="sc-dnqmqq qfvou"
      data-type="add-button"
      type="submit"
    >
      <svg
        aria-hidden="true"
        className="StyledIconBase-sc-bdy9j4 jbPeIR sc-htpNat dPFygr"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </button>
  </form>,
  <section
    className="sc-bZQynM dxNOjd"
  />,
]
`;

exports[`First Snapshot test should render TodoForm correctly 1`] = `ShallowWrapper {}`;

exports[`First Snapshot test should render header correctly 1`] = `ShallowWrapper {}`;
